### ISSUE TITLE
Define developer-facing api for runtime configuration and introspection, hide risky object properties.

### ISSUE THREAD

**Issue by aeschylus** [ProjectMirador/mirador#574]
It is currently not well documented and not well supported to listen to Mirador's events, respond to them reliably, and change object properties at runtime in a safe way. 

More enterprising deployers must read the source code to customise behaviour. What kinds of introspection, event hijacking, or other customisation or dynamic/scriptable features would the community like to have in Mirador? 

Is it/will it be sufficient to simply better document the current methods that may be useful to developers with an interest in customisation?

**Comment by aeschylus** [comment]
See https://github.com/IIIF/mirador/issues/573

**Comment by noahadler** [comment]
I'm not sure if this should be attached to this issue or be its own, but we have a use case that I expect may be fairly common, especially for folks trying to integrate Mirador into a reactive UI framework (Angular, React, etc.)  These frameworks add and remove elements from the DOM on a regular basis, but it seems that Mirador will hang around even if the parent DOM element it's bound to is removed by the reactive framework.

After chatting in Gitter, it's my understanding that there is currently no prescribed way to destroy or clean up Mirador after instantiation.  My cursory initial tests indicate there are a fair number of unpredictable handles, such as closure variables, detached DOM nodes, jQuery events bound to the document, that keep Mirador instantiations from being garbage collected.  I don't have the depth of familiarity to know what all might be involved with this, but would like to request some kind of official approach be developed.

tl;dr - requesting an official API for destroying Mirador instances, while remaining on the same browser page

Thanks!

**Comment by beaudet** [comment]
The ConservationSpace project needs access to the following Mirador data structures / methods:
* callback when a canvas is selected so that the active canvas and its manifest can be inspected by an external application - if canvas has one image, we intend to highlight that image in a conservation space dialog
* callback when an annotation region is created, selected, or deleted
* ability to create, select, and delete annotations via API calls
* mechanism to provide alternative annotation editing dialog
* API identified to highlight one or more annotations and the corresponding annotation regions
* ability to append / remove a canvas from a manifest
* API to lock / unlock all UI controls
* API to switch whether user can create and / or edit annotations

**Comment by hotnuts21** [comment]
+1 for destroying an instance of Mirador. For us to enable in page language switching of the interface via a lang variable.

**Comment by aeschylus** [comment]
See also #904, #583

**Comment by camillevilla** [comment]
This ticket is being closed as part of the Mirador 3 issues review. We encourage all Mirador 2 implementers to try out Mirador 3 and leave feedback in new tickets.
